### PR TITLE
tests: add UI tests (SDKCF-2246)

### PR DESCRIPTION
**Pull request by imaginaris**
# Description
Added basic UI tests.
Any ideas for more tests or for any improvements are welcome.

## Links
SDKCF-2246

# Checklist
- [X] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [X] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data **before every commit**, including API endpoints and keys
- [X] I ran `fastlane ci` without errors

### CODE REVIEW

File: UITests/SlideUpViewSpec.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    launchAppIfNecessary(args: "-campaignType slide-up-redirect")
+                    if !iamView.exists {
+                        app.buttons["login_successful"].tap()
+                        expect(iamView.exists).toEventually(beTrue())
+                    }
+                }
+
+                afterEach {
+                    app.activate() // returns from the browser
+                    usleep(250000) // wait for app switch animation to finish
```
Comment (apparition47) [inline UITests/SlideUpViewSpec.swift]: Super long delay is a good catch-all in terms of reliability but I personally prefer either disabling animations (or speeding them up at least) just to make UI tests faster and more deterministic

File: UITests/Campaigns/modal-controls.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        "header": "Test Campaign",
+        "titleColor": "#000000",
+        "headerColor": "#000000",
+        "messageBodyColor": "#000000",
+        "backgroundColor": "#ffffff",
+        "frameColor": "#ffffff",
+        "resource":
+        {
+            "assetsUrl": "",
+            "imageUrl": "file:///System/Library/CoreServices/CoreTypes.bundle/Contents/Resources/ClippingPicture.icns",
```
Comment (apparition47) [inline UITests/Campaigns/modal-controls.json]: Think this file will be around in future versions of macOS? 🤔 Or what if we execute tests on a real iOS device?

File: UITests/FullScreenViewSpec.swift [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                    }
+                }
+
+                it("should not display any buttons") {
+                    expect(iamView.buttons["Button0"].exists).to(beFalse())
+                    expect(iamView.buttons["Button1"].exists).to(beFalse())
+                }
+
+                it("should not display opt-out checkbox") {
+                    expect(iamView.buttons["Do not show me this message again"].exists).to(beFalse())
```
Comment (apparition47) [inline UITests/FullScreenViewSpec.swift]: Will this work if the device is set to Japanese or are we forcing EN locale for our test env?

File: UITests/SlideUpViewSpec.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
+import XCTest
```
Comment (apparition47) [inline UITests/SlideUpViewSpec.swift]: `Quick` already imports `XCTest`

**Review by apparition47 — CHANGES_REQUESTED** [review]

File: UITests/SlideUpViewSpec.swift (hunk 4 above)
Comment (imaginaris) [reply UITests/SlideUpViewSpec.swift]: You are right!

File: UITests/FullScreenViewSpec.swift (hunk 3 above)
Comment (imaginaris) [reply UITests/FullScreenViewSpec.swift]: Tests always run in english. But actually, I can add an identifier and use it instead

File: UITests/Campaigns/modal-controls.json (hunk 2 above)
Comment (imaginaris) [reply UITests/Campaigns/modal-controls.json]: Yeah, might not be. On real device it'll probably fail but are we ever going to run those tests on a real device?
I didn't want to put any http url based image. I can try to embed some image in the test bundle but the path will be different on real device as well. Do you have any better idea?

File: UITests/SlideUpViewSpec.swift (hunk 1 above)
Comment (imaginaris) [reply UITests/SlideUpViewSpec.swift]: But that's just 0.25s 😞 It's for the system's app switch animation - I don't have any control over it. I know it's not a pretty solution but I could't find any other way... Without this the test runner tries to tap a button when it still "moving"
```
Failed to synthesize event: Failed to scroll to visible (by AX action) Button, label: 'custom_test', error: Error kAXErrorCannotComplete performing AXAction 2003 on element AX element pid: 98091, elementOrHash.elementID: 140614030137392.10. (Underlying Error: Error kAXErrorCannotComplete performing AXAction 2003 on element AX element pid: 98091, elementOrHash.elementID: 140614030137392.10)
```

Comment (imaginaris) [reply UITests/SlideUpViewSpec.swift]: I'm removing those tests that involve app transition (opening url). They are too unstable

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift [hunk 5]
```diff
@@ -15,12 +17,12 @@ struct AtomicGetSet<Value> {
             return queue.sync { value }
         }
         set {
-            queue.sync { value = newValue }
+            queue.sync(flags: .barrier) { value = newValue }
```
Comment (imaginaris) [inline RInAppMessaging/Classes/Utilities/AtomicWrapper.swift]: changing `AtomicGetSet` to class and using concurrent queue fixes intermittent crashes that were happening (I don't fully understand why though)

File: fastlane/Fastfile [hunk 6]
```diff
@@ -14,10 +14,25 @@ platform :ios do
...
+      slack_only_on_failure: true,
+      skip_slack: ENV['PREVIOUS_BUILD_STATUS'] == 'error', # var defined in Bitrise step
+      output_types: 'json-compilation-database,html,junit',
+      output_files: 'compile_commands.json,report.html,report.junit')
+
+    scan(
+      clean: false,
+      skip_build: true,
       output_directory: './artifacts/unit-tests',
       scheme: ENV['REM_FL_TESTS_SCHEME'] || 'Tests',
```
Comment (imaginaris) [inline fastlane/Fastfile:31]: unit test must be run as the last step in order to generate code coverage report for Danger (we could probably add some custom paths for the reports instead)

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (apparition47) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift]: Do you have a stacktrace of the crash(es)? It looks like you've seen [Donny Wals' suggestions](https://www.donnywals.com/why-your-atomic-property-wrapper-doesnt-work-for-collection-types/) to fix the problem. But it still seems our implementation might not do so well with concurrency and collection value-types especially when collection value types are copied by each thread and copies also given out in the property wrapper's `wrappedValue.get` like he mentions

File: UITests/Campaigns/modal-controls.json (hunk 2 above)
Comment (donnie-jp) [reply UITests/Campaigns/modal-controls.json]: Can it use an image embedded as a bundle resource?

File: UITests/SlideUpViewSpec.swift (hunk 1 above)
Comment (apparition47) [reply UITests/SlideUpViewSpec.swift]: For the system app switching transitions, does the "Reduce Motion" accessibility setting make a difference?

File: Podfile [hunk 7]
```diff
@@ -7,15 +7,28 @@ target 'RInAppMessaging_Example' do
...
+    end
   end
 end
 
 post_install do |installer|
   system("./configure-secrets.sh InAppMessaging #{secrets.join(" ")}")
+  installer.pods_project.targets.each do |target|
+    target.build_configurations.each do |config|
+      if target.name == 'RInAppMessaging'
+        config.build_settings['ENABLE_TESTABILITY'] = 'YES'
```
Comment (apparition47) [inline Podfile]: What's this for? [`ENABLE_TESTABILITY=YES` is already set for the parent `RInAppMessaging` project](https://github.com/rakutentech/ios-inappmessaging/blob/master/RInAppMessaging.xcodeproj/project.pbxproj#L702) then it should be inherited by your new `UITest` target

Comment (imaginaris) [reply Podfile]: RInAppMessaging target inherits from `Pods` project unfortunately.

File: RInAppMessaging/Classes/Helpers/UITestHelper.swift [hunk 8]
```diff
@@ -0,0 +1,32 @@
+internal struct UITestHelper {
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Helpers/UITestHelper.swift]: I'm not keen on this being in the SDK folder's source.

File: UITests/Campaigns/modal-controls.json (hunk 2 above)
Comment (donnie-jp) [reply UITests/Campaigns/modal-controls.json]: Bump

File: fastlane/Fastfile [hunk 9]
```diff
@@ -14,10 +14,25 @@ platform :ios do
...
-      device: ENV['REM_FL_TESTS_DEVICE'] || 'REM SDK Unit Tests',
+      device: ENV['REM_FL_TESTS_DEVICE'] || 'iPhone 11',
```
Comment (donnie-jp) [inline fastlane/Fastfile:32]: 👍

**Review by donnie-jp — CHANGES_REQUESTED** [review]
I really like that we're adding UI tests. Nice work 👍 

However, I'm not so comfortable with us interweaving the "testing environment" checks through our SDK code. I'd like to have a more seamless way of achieving that and with clearer separation between real and test implementation. Could we achieve it through overriding or possibly using protocols (with default implementation in SDK code and test-specific one in UI test target)?

File: UITests/SlideUpViewSpec.swift (hunk 1 above)
Comment (imaginaris) [reply UITests/SlideUpViewSpec.swift]: I haven't checked that but I have a feeling that it wouldn't help much

File: UITests/Campaigns/modal-controls.json (hunk 2 above)
Comment (imaginaris) [reply UITests/Campaigns/modal-controls.json]: Probably. If putting image name as a url is going to work...
But yeah, it is a better idea

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift]: @apparition47 That's why I tried to combine his AtomicDict and property wrapper which resulting in changing struct to class and adding concurrent queue. It seems like it helps but to be honest I'm not sure what is what at this point. (I'd have to calmly analyse this case one more time).
Here's the stacktrace (crashed on current line):
```
* thread #20, queue = 'IAM.AtomicProperty489)', stop reason = EXC_BAD_ACCESS (code=1, address=0x3be9ea32eab0)
    frame #0: 0x000000010924f2bd libswiftCore.dylib`_swift_release_dealloc + 13
    frame #1: 0x0000000108c0a0bd RInAppMessaging`partial apply for specialized closure #1 in AtomicGetSet.wrappedValue.setter [inlined] generic not re-abstracted specialization <Swift.Dictionary<Swift.String, RInAppMessaging.UserDataCacheContainer>> of closure #1 () -> () in RInAppMessaging.AtomicGetSet.wrappedValue.setter : τ_0_0 at <compiler-generated>:0 [opt]
    frame #2: 0x0000000108c0a09e RInAppMessaging`partial apply for specialized closure #1 in AtomicGetSet.wrappedValue.setter at <compiler-generated>:0 [opt]
    frame #3: 0x0000000108be5dcc RInAppMessaging`thunk for @callee_guaranteed () -> () at <compiler-generated>:0 [opt]
    frame #4: 0x0000000108be5dee RInAppMessaging`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0 [opt]
    frame #5: 0x00007fff201078df libdispatch.dylib`_dispatch_client_callout + 8
    frame #6: 0x00007fff201155e1 libdispatch.dylib`_dispatch_lane_barrier_sync_invoke_and_complete + 94
  * frame #7: 0x0000000108c08e97 RInAppMessaging`specialized AtomicGetSet.wrappedValue.setter(newValue=<unavailable>, self=<unavailable>) at AtomicWrapper.swift:20:19 [opt]
    frame #8: 0x0000000108c62caf RInAppMessaging`UserDataCache.cacheCampaignData(data=0 values, userIdentifiers=<unavailable>, self=0x0000600000786980) at UserDataCache.swift:55:36 [opt]
    frame #9: 0x0000000109e97d94 Tests`merged closure #1 () -> () in closure #1 () -> () in closure #8 () -> () in closure #2 () -> () in Tests.UserDataCacheSpec.spec() -> () + 212
    frame #10: 0x0000000109e97c7e Tests`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0 [opt]
    frame #11: 0x00007fff2010670d libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #12: 0x00007fff201078df libdispatch.dylib`_dispatch_client_callout + 8
    frame #13: 0x00007fff2010de15 libdispatch.dylib`_dispatch_lane_serial_drain + 715
    frame #14: 0x00007fff2010e98c libdispatch.dylib`_dispatch_lane_invoke + 400
    frame #15: 0x00007fff20118f81 libdispatch.dylib`_dispatch_workloop_worker_thread + 772
    frame #16: 0x00007fff6034045d libsystem_pthread.dylib`_pthread_wqthread + 314
    frame #17: 0x00007fff6033f42f libsystem_pthread.dylib`start_wqthread + 15
```

File: RInAppMessaging/Classes/Helpers/UITestHelper.swift (hunk 8 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Helpers/UITestHelper.swift]: I know but to be honest I think this is the least intrusive way of configuring the SDK for UI Tests.
We could use some external tool to mock the endpoints but that's much more complex solution.

**Comment by imaginaris** [comment]
> I really like that we're adding UI tests. Nice work 👍
> 
> However, I'm not so comfortable with us interweaving the "testing environment" checks through our SDK code. I'd like to have a more seamless way of achieving that and with clearer separation between real and test implementation. Could we achieve it through overriding or possibly using protocols (with default implementation in SDK code and test-specific one in UI test target)?

@donnie-jp But in UITests you can't access module's internal classes. I can only use public API which goes against the idea of performing UI Tests.

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (donnie-jp) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift]: @imaginaris I'd like this change considered separately. Please extract the AtomicWrapper related changes to another PR and ensure there is a JIRA tracking it. 

I think this change before/after is a good candidate for the automated stress tests we are planning.

File: RInAppMessaging/Classes/Helpers/UITestHelper.swift (hunk 8 above)
Comment (donnie-jp) [reply RInAppMessaging/Classes/Helpers/UITestHelper.swift]: I think it's a common pattern on Android to use MockWebServer for running (UI) tests and on iOS there's the possibility of using https://github.com/httpswift/swifter - could be a better approach for us?

Comment (donnie-jp) [reply RInAppMessaging/Classes/Helpers/UITestHelper.swift]: I'd be more comfortable if this struct (or the body/contents of the struct preferably) was namespaced under `Environment` enum as `UITesting`

**Review by donnie-jp — CHANGES_REQUESTED** [review]
I think this will be good to go after these tasks are done:
- moving UITestHelper stuff under the Environment namespace
- extracting the AtomicWrapper changes into a separate PR

File: RInAppMessaging/Classes/Utilities/AtomicWrapper.swift (hunk 5 above)
Comment (malayli) [reply RInAppMessaging/Classes/Utilities/AtomicWrapper.swift]: I think that this change should also be done in RAnalytics too as it uses AtomicGetSet.

**Review by malayli — APPROVED** [review]

File: UITests/Campaigns/modal-controls.json (hunk 2 above)
Comment (imaginaris) [reply UITests/Campaigns/modal-controls.json]: There's no way to get a static url to a resource :( I used Shock url mocking instead

File: RInAppMessaging/Classes/Helpers/UITestHelper.swift (hunk 8 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Helpers/UITestHelper.swift]: I removed `UITestHelper` completely. The SDK is now entirely UITest-agnostic

**Comment by imaginaris** [comment]
@donnie-jp @malayli @apparition47 
I made some final updates to this PR. I replaced backend mock injection logic with third party solution called Shock.
Thanks to that, UITests are now non intrusive and the SDK is not aware of any new environment.
Please review latest commit whenever you have some time.

**Review by donnie-jp — APPROVED** [review]
Nice!
It's great you found a way to not pollute the SDK classes with test-specific code and the Shock mock server looks very useful. 👍

**Review by apparition47 — APPROVED** [review]
Great!

**Pull request merged**